File: src/main/java/com/jme/asset/store/db/entity/asset/AssetCategoryEntity.java
package com.jme.asset.store.db.entity.asset;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.jme.asset.store.db.entity.BaseEntity;

import org.jetbrains.annotations.Nullable;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;

/**
 * The asset category entity
 *
 * @author Yunkevich Andrei
 */
@Entity
@Table(name = "ASSET_CATEGORY_ENTITY")
public class AssetCategoryEntity extends BaseEntity {

    /**
     * The name of asset category
     */
    @Column(name = "name")
    private String name;

    /**
     * The description of asset category
     */
    @Column(name = "description")
    private String description;

    /**
     * The parent of asset category
     */
    @ManyToOne(cascade = {CascadeType.ALL})
    @JoinColumn(name = "parent_id")
    @JsonBackReference
    private AssetCategoryEntity parent;

    /**
     * The children of asset category
     */
    @OneToMany(mappedBy = "parent")
    @JsonManagedReference
    private Set<AssetCategoryEntity> children = new HashSet<AssetCategoryEntity>();

    public AssetCategoryEntity() {
    }

    /**
     * Get the category name
     *
     * @return the name of asset category
     */
    public @Nullable String getName() {
        return name;
    }

    /**
     * Set the category name
     *
     * @param name the name of asset category
     */
    public void setName(@Nullable final String name) {
        this.name = name;
    }

    /**
     * Get the description of category
     *
     * @return the description of asset category
     */
    public @Nullable String getDescription() {
        return description;
    }

    /**
     * Set the category description
     *
     * @param description the description of asset category
     */
    public void setDescription(@Nullable final String description) {
        this.description = description;
    }

    /**
     * Get the parent of asset category
     *
     * @return the parent of asset category
     */
    public @Nullable AssetCategoryEntity getParent() {
        return parent;
    }

    /**
     * Set the category parent
     *
     * @param parent the parent of asset category
     */
    public void setParent(@Nullable final AssetCategoryEntity parent) {
        this.parent = parent;
    }

    /**
     * Get the children of asset category
     *
     * @return the children of asset category
     */
    public @Nullable Set<AssetCategoryEntity> getChildren() {
        return children;
    }

    /**
     * Set the category children
     *
     * @param children the children of asset category
     */
    public void setChildren(@Nullable final Set<AssetCategoryEntity> children) {
        this.children = children;
    }
}

File: front-end/src/model/entity/AssetCategoryEntity.ts
/**
 * The asset category entity.
 *
 * @author Denis Lesheniuk.
 */
export class AssetCategoryEntity {

  /**
   * The id of asset category
   */
  private _id: number;

  /**
   * The name of asset category
   */
  private _name: string;

  /**
   * The description of asset category
   */
  private _description: string;

  /**
   * The parent of asset category
   */
  private _parent: AssetCategoryEntity;

  /**
   * The children of asset category
   */
  private _children: AssetCategoryEntity[];

  constructor() {
  }

  /**
   * Get the asset component id.
   *
   * @returns {number} the asset component id.
   */
  get id(): number {
    return this._id;
  }

  /**
   * Set the asset component id.
   *
   * @param {number} value the asset component id.
   */
  set id(value: number) {
    this._id = value;
  }

  /**
   * Get the asset category name;
   *
   * @returns {string} the asset category name.
   */
  get name(): string {
    return this._name;
  }

  /**
   * Set the asset category name.
   *
   * @param {string} value the casset category name.
   */
  set name(value: string) {
    this._name = value;
  }

  /**
   * Get the asset category description.
   *
   * @returns {string} the asset category description.
   */
  get description(): string {
    return this._description;
  }

  /**
   * Set the asset category description.
   *
   * @param {string} value the asset category description.
   */
  set description(value: string) {
    this._description = value;
  }

  /**
   * Get the asset category parent.
   *
   * @returns {AssetCategoryEntity} the asset category parent.
   */
  get parent(): AssetCategoryEntity {
    return this._parent;
  }

  /**
   * Set the asset category parent.
   *
   * @param {AssetCategoryEntity} value
   */
  set parent(value: AssetCategoryEntity) {
    this._parent = value;
  }

  /**
   * Get the asset category children.
   *
   * @returns {AssetCategoryEntity[]} the asset category children.
   */
  get children(): AssetCategoryEntity[] {
    return this._children;
  }

  /**
   * Set the asset category children.
   *
   * @param {AssetCategoryEntity[]} value the asset category children.
   */
  set children(value: AssetCategoryEntity[]) {
    this._children = value;
  }
}
